crates/bevy_reflect/src/path/parse.rs

```diff
@@ -0,0 +1,173 @@
+    CloseBeforeOpen,
+}
+
+pub(super) struct PathParser<'a> {
+    path: &'a str,
+    offset: usize,
+}
+impl<'a> PathParser<'a> {
+    pub(super) fn new(path: &'a str) -> Self {
+        PathParser { path, offset: 0 }
+    }
+
+    fn next_token(&mut self) -> Option<Token<'a>> {
+        let input = &self.path[self.offset..];
+
+        // Return with `None` if empty.
+        let first_char = input.chars().next()?;
+
+        if let Some(token) = Token::symbol_from_char(first_char) {
+            self.offset += 1; // NOTE: we assume all symbols are ASCII
+            return Some(token);
+        }
+        // We are parsing either `0123` or `field`.
+        // If we do not find a subsequent token, we are at the end of the parse string.
+        let ident = input.split_once(Token::SYMBOLS).map_or(input, |t| t.0);
+
+        self.offset += ident.len();
+        Some(Token::Ident(Ident(ident)))
+    }
+
+    fn next_ident(&mut self) -> Result<Ident<'a>, Error<'a>> {
+        match self.next_token() {
+            Some(Token::Ident(ident)) => Ok(ident),
+            Some(other) => Err(Error::ExpectedIdent(other)),
+            None => Err(Error::NoIdent),
+        }
+    }
+
+    fn access_following(&mut self, token: Token<'a>) -> Result<Access<'a>, Error<'a>> {
+        match token {
+            Token::Dot => Ok(self.next_ident()?.field()),
+            Token::CrossHatch => self.next_ident()?.field_index(),
+            Token::Ident(ident) => Ok(ident.field()),
+            Token::CloseBracket => Err(Error::CloseBeforeOpen),
+            Token::OpenBracket => {
+                let index_ident = self.next_ident()?.list_index()?;
+                match self.next_token() {
+                    Some(Token::CloseBracket) => Ok(index_ident),
+                    Some(other) => Err(Error::BadClose(other)),
+                    None => Err(Error::Unclosed),
+                }
+            }
+        }
+    }
+}
+impl<'a> Iterator for PathParser<'a> {
+    type Item = (Result<Access<'a>, ReflectPathError<'a>>, usize);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let token = self.next_token()?;
+        let offset = self.offset;
+        let err = |error| ReflectPathError::ParseError {
+            offset,
+            path: self.path,
+            error: ParseError(error),
+        };
+        Some((self.access_following(token).map_err(err), offset))
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct Ident<'a>(&'a str);
+
+impl<'a> Ident<'a> {
+    fn field(self) -> Access<'a> {
+        let field = |_| Access::Field(self.0.into());
+        self.0.parse().map(Access::TupleIndex).unwrap_or_else(field)
+    }
+    fn field_index(self) -> Result<Access<'a>, Error<'a>> {
+        Ok(Access::FieldIndex(self.0.parse()?))
+    }
+    fn list_index(self) -> Result<Access<'a>, Error<'a>> {
+        Ok(Access::ListIndex(self.0.parse()?))
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum Token<'a> {
+    Dot,
+    CrossHatch,
```

Comment:
Not used to that word for `#`, son uses `Pound`: https://docs.rs/syn/latest/syn/token/struct.Pound.html

Reply:
I would prefer using `Pound`. It's a private type so no breaking change. I was just re-using the term as it was before the refactor.